app/domain/README.md

```diff
@@ -0,0 +1,296 @@
+many gems make it easier to write value objects and command objects.
+
+In general, there is a tradeoff when using gems:
+
+- Positive: 
+  - They cut down on boilerplate and make your code more concise
+- Negative: 
+  - Every new developer has another extra thing to learn
+  - The code is often not as obvious as plain ruby
+
+You'll encounter two gems when working in Conjur:
+
+- `command_class` for writing command objects
+  - This is still used extensively.  Read below for more information.
+- `dry-struct` for value objects
+  - While this is a good gem, I've stopped using inside Conjur for the reasons
+    listed above
+  - I now write value objects using plain ruby
+
+## `CommandClass` Deep Dive
+
+The [`command_class` README](https://github.com/jonahx/command_class) has basic
+information about usage.  Here I'll note points of confusion and design
+considerations.
+
+### Dependencies vs Inputs
+
+Remember that an instance of a `CommandClass` is just a ruby proc, as
+is anything with a `call` method.  Roughly, then, you can think of a proc as a
+function.  So a `CommandClass` is really just a _configurable function_, and an
+instance is a _configured function_.
+
+How is it configured?  With its dependencies.
+
+The _inputs_, by contrast, are the _configured function's arguments_.  They're
+what you pass when you call the "function".
+
+Another way to think of this is that dependencies are known when the program
+starts: Your entire program, even a multi-threaded one, needs only a single
+instance of each `CommandClass` it uses.  In Conjur, we often create
+`CommandClass` instances in controllers on each request. This avoids dealing
+with Rails initializers, but conceptually speaking it's an abuse of the
+pattern, even though it doesn't matter much practically.
+
+The inputs, on the other hand, are known only at runtime. They're often, but
+not necessarily, given as the params of an http request.  
+
+### Pitfalls To Avoid
+
+- Hardcoding dependencies that should be injected.
+  - For example: Directly referencing `Rails.logger` or a model class.
+  - All code with side effects should be injected as a depenency.
+- Names that are too long or require additional comments to explain.
+  - Too verbose: `PerformValidationChecks`.  Better:  `ValidateNewUser`
+
+### Design is still paramount
+
+Explicit dependencies help you write code that's easy to test and read, but
+they don't guarantee it.  It's still possible to write confusing or complex
+code using these patterns.  There is still some art to good design that's hard
+to capture.  Nothing beats a set of fresh, critical eyes to validate
+readability.
+
+A few heuristics to keep in mind:
+
+- Avoid putting too many responsibilities into a command class.  Its single
+  verb phrase name should naturally describe it.
+  - Split up a `CommandClass` when needed
+- Avoid putting too little in one.  Not everything should be a `CommandClass`.
+  Keep in mind why we write them.
```

Comment:
i'd like to get an example here as this is something that i never really understood. From my point of view, if you have a responsibility that needs to be done in several places then it makes sense to make it a command class. i have an example in another PR which i will link to once it is merged.

Reply:
here is the example. we have this class: https://github.com/cyberark/conjur/blob/master/app/domain/authentication/authn_k8s/extract_container_name.rb

it is used in both [here](https://github.com/cyberark/conjur/blob/master/app/domain/authentication/authn_k8s/inject_client_cert.rb#L177) and [here](https://github.com/cyberark/conjur/blob/master/app/domain/authentication/authn_k8s/validate_pod_request.rb#L88)

you will see that the command class has only a dependency on `Rails.logger` which may point that it should not be a command class. However, it does have some logic and constants that i think should not be implemented in both places. 

i'd like to get your thoughts on this, and even more important - extract a rule of thumb for this README.

Reply:
This is a good question, and it's come a few places in your PRs as well as others.  I'll try a short-ish answer here.  Tell me what you think, or if it raises other questions....

> From my point of view, if you have a responsibility that needs to be done in several places then it makes sense to make it a command class.

This is mostly try if:

1. The logic has side-effects.
2. The logic has minimum complexity

The reason I add the 2. caveat is if it's really simple, like 1 line, it might be simpler to repeat it.  Again, introducing an abstraction is not cost-free.  Just like it's possible to have too many tiny methods in a class, it's possible to have too many tiny command classes... you may have removed duplication, but you force the reader to know about more things, and to be able to remember and use them properly.  This is a deep subject, and I think the best way to get a handle on it personally and as a team is to have discussions about multiple specific examples.

> here is the example. we have this class: https://github.com/cyberark/conjur/blob/master/app/domain/authentication/authn_k8s/extract_container_name.rb

I remember seeing this one before. 

So I think my take on this one is that we're missing two things:

1. A value object like `K8sHost`, which is either created with `service_id` and has a `container_name` method, or is created without a `service_id` and has a `container_name(service_id)` method, depending on your needs.
2. A repository object which would return a `K8sHost`, given a `host_id` (or given a `host_id`/`service_id` combo)

We don't use `Repo` objects formally in Conjur, though I have may have created 1 or 2, and we treat the Rails model objects like repos in many places.  But the basic idea is that the repository knows all the database storage details, and it just give you back a nice, clean application object that works the way you want it to.

So in this particular case,  all this junk:

```ruby
 def container_name
        annotation_value("authn-k8s/#{@service_id}/#{AUTHENTICATION_CONTAINER_NAME_ANNOTATION}") ||
          annotation_value("authn-k8s/#{AUTHENTICATION_CONTAINER_NAME_ANNOTATION}") ||
          annotation_value("kubernetes/#{AUTHENTICATION_CONTAINER_NAME_ANNOTATION}") ||
          default_authentication_container_name
      end
```
is just "details about how Conjur happens to store k8s info in its database using annotations".  Your application code does't care about that crap.  That's just database implementation detail.  So we want our Repo, and _only_ our repo, to know about this stuff.  We'd just query it and get back `K8sHost`, say.  

So this stuff fits more naturally at the at lower level than in a `CommandClass` object, which (per this README) are intended for the higher level coordination of multiple tasks into a single use-case.